firestore/src/common/local_cache_settings.cc

```diff
@@ -0,0 +1,138 @@
+  return new_settings;
+}
+
+int64_t PersistentCacheSettings::size_bytes() const {
+  return settings_internal_->size_bytes();
+}
+
+const LocalCacheSettingsInternal& PersistentCacheSettings::internal() const {
+  return *settings_internal_;
+}
+
+MemoryEagerGCSettings MemoryEagerGCSettings::Create() { return {}; }
+
+MemoryEagerGCSettings::MemoryEagerGCSettings() {
+  settings_internal_ = std::make_shared<MemoryEagerGCSettingsInternal>();
+}
+
+const MemoryGarbageCollectorSettingsInternal& MemoryEagerGCSettings::internal()
+    const {
+  return *settings_internal_;
+}
+
+MemoryLruGCSettings MemoryLruGCSettings::Create() { return {}; }
+
+MemoryLruGCSettings::MemoryLruGCSettings() {
+  settings_internal_ = std::make_shared<MemoryLruGCSettingsInternal>();
+}
+
+MemoryLruGCSettings::MemoryLruGCSettings(
+    const MemoryLruGCSettingsInternal& other) {
+  settings_internal_ = std::make_shared<MemoryLruGCSettingsInternal>(other);
+}
+
+MemoryLruGCSettings MemoryLruGCSettings::WithSizeBytes(int64_t size) {
+  MemoryLruGCSettings result;
+  result.settings_internal_ = std::make_shared<MemoryLruGCSettingsInternal>(
+      this->settings_internal_->WithSizeBytes(size));
+  return result;
+}
+
+int64_t MemoryLruGCSettings::size_bytes() const {
+  return settings_internal_->size_bytes();
+}
+
+const MemoryGarbageCollectorSettingsInternal& MemoryLruGCSettings::internal()
+    const {
+  return *settings_internal_;
+}
+
+MemoryCacheSettings MemoryCacheSettings::Create() { return {}; }
+
+MemoryCacheSettings::MemoryCacheSettings() {
+  settings_internal_ = std::make_shared<MemoryCacheSettingsInternal>();
+}
+
+MemoryCacheSettings MemoryCacheSettings::WithGarbageCollectorSettings(
+    const MemoryGarbageCollectorSettings& settings) const {
+  MemoryCacheSettings result;
+  result.settings_internal_ = std::make_shared<MemoryCacheSettingsInternal>(
+      this->settings_internal_->WithGarbageCollectorSettings(settings));
+  return result;
+}
+
+const LocalCacheSettingsInternal& MemoryCacheSettings::internal() const {
+  return *settings_internal_;
+}
+
+bool operator==(const LocalCacheSettings& lhs, const LocalCacheSettings& rhs) {
+  return lhs.kind() == rhs.kind() && lhs.internal() == rhs.internal();
+}
+
+bool operator==(const MemoryCacheSettings& lhs,
+                const MemoryCacheSettings& rhs) {
+  return &lhs == &rhs || (*lhs.settings_internal_ == *rhs.settings_internal_);
+}
+
+bool operator==(const PersistentCacheSettings& lhs,
+                const PersistentCacheSettings& rhs) {
+  return &lhs == &rhs || (*lhs.settings_internal_ == *rhs.settings_internal_);
+}
+
+bool operator==(const MemoryGarbageCollectorSettings& lhs,
+                const MemoryGarbageCollectorSettings& rhs) {
+  return lhs.internal() == rhs.internal();
+}
+
+bool operator==(const MemoryEagerGCSettings& lhs,
+                const MemoryEagerGCSettings& rhs) {
+  return &lhs == &rhs || (*lhs.settings_internal_ == *rhs.settings_internal_);
```

Comment:
Is there a compelling reason to optimize `operator==()` with `&lhs == &rhs`? I can see it as a performance optimization, but I don't think that the performance benefit outweighs the detraction from consistency with the other `operator==` overloads. Also applies to the `operator==()` for `MemoryLruGCSettings`.